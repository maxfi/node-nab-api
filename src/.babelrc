{
  "env": {
    "production": {
      "presets": [
        [
          "@babel/preset-env",
          {
            "targets": {
              "node": "4"
            },
            "modules": false,
            "useBuiltIns": "usage",
            "debug": true
          },
          "@babel/preset-flow"
        ]
      ],
      "plugins": ["@babel/plugin-transform-runtime"]
    },
    "development": {
      "presets": [
        [
          "@babel/preset-env",
          {
            "useBuiltIns": "usage"
          },
          "@babel/preset-flow"
        ]
      ]
    }
  },
  "plugins": ["@babel/plugin-proposal-object-rest-spread"]
}
