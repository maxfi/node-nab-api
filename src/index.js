import { js2xml, xml2js } from 'xml-js'
import fetch from 'node-fetch'
import getMessageContainer from './util/get-message-container'
import getResponse from './util/get-response'
import { normaliseDateString } from './util/normalise-date-string'

/**
 * NAB payment gateway API wrapper for node.js
 * @module node-nab-api
 */

const toXml = x => js2xml(x, { spaces: 4, compact: true })
const fromXml = x => xml2js(x, { compact: true })

/**
 * Make the request to the NAB payment gateway API.
 * Converts the payload to XML and converts the response to Object.
 * @private
 * @param {String} url - Endpoint URL.
 * @param {Object} payload - Request body.
 * @return {Object} Returns the API response as Object.
 */
const request = async (url, payload) => {
  const xml = toXml(payload)
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'text/xml' },
    body: xml,
  })
  if (!response.ok) {
    const { status, statusText } = response
    console.log('REQUEST:', url, ':', xml)
    throw new Error(`Bad response from NAB API: ${status} - ${statusText}`)
  }
  const body = await response.text()
  return getResponse(fromXml(body))
}

/**
 * Extracts the relevant data from the response
 * @private
 * @param {Object} response - API response as Object
 * @param {String} el - Name of the root XML element that contains the data
 * @return {Object} {MessageInfo, RequestType, MerchantInfo, Status, Data}
 */
const clean = (response, el) => {
  const result = response.NABTransactMessage
  result.Data = result[el][el + 'List'][el + 'Item']
  delete result.Data._attributes
  delete result[el]
  return result
}

/**
 * @class NAB
 */
export default class NAB {
  /**
   * Creates an instance of the NAB class.
   * @constructor
   * @param {Object} config Config.
   * @param {String} config.merchantId 5 character merchant ID supplied by NAB Transact.
   * @param {String} config.password Transaction password. Password used for authentication of the merchant's request message, supplied by NAB Transact.
   * @param {Object} [options] Options.
   * @param {Boolean} options.testMode Set to `true` to direct requests to the NAB payment gateway test environment.
   * @param {String} options.timeout Timeout value used, in seconds.
   */
  constructor(config, options = {}) {
    this.merchantId = config.merchantId
    this.password = config.password
    this.baseUrl = options.testMode
      ? 'https://demo.transact.nab.com.au/xmlapi/'
      : 'https://transact.nab.com.au/live/xmlapi/'
  }

  _getCredentials() {
    return {
      merchantId: this.merchantId,
      password: this.password,
    }
  }

  async _post(type, payload) {
    // Capitalize the element. Eg -> 'token' to 'Token'
    const el = type.charAt(0).toUpperCase() + type.slice(1)
    const url = this.baseUrl + type
    const response = await request(url, payload)
    return clean(response, el)
  }

  /**
   * @summary Add a new token
   * @param {String} messageId The `messageId` is a reference for the xml request. If you had a server internally you could store all your xml request and this would be a way to locate the request if a payment was to fail. A request will still be sent if they have the same `messageId` and will be treated as a new xml request.
   * @param {Object} tokenDetails Token details.
   * @param {String} tokenDetails.cardNumber Credit card number.
   * @param {String} tokenDetails.expiryDate Credit card expiry date.
   * @returns {Promise} {MessageInfo, RequestType, MerchantInfo, Status, Data}
   */
  addToken(messageId, tokenDetails) {
    const payload = getMessageContainer({
      messageId,
      credentials: this._getCredentials(),
      requestType: 'addToken',
      dataElementName: 'Token',
      payload: {
        ...tokenDetails,
        tokenType: 1, // The type of token created. Defaults to 1 if absent. Type 1 is 16 digits, not based on the card number, failing the LUHN check.
      },
    })
    return this._post('token', payload)
  }

  /**
   * @param {String} messageId The `messageId` is a reference for the xml request. If you had a server internally you could store all your xml request and this would be a way to locate the request if a payment was to fail. A request will still be sent if they have the same `messageId` and will be treated as a new xml request.
   * @param {String} tokenValue The token value that represents a stored card within NAB Transact.
   * @returns {Promise} {MessageInfo, RequestType, MerchantInfo, Status, Data}
   */
  lookupToken(messageId, tokenValue) {
    const payload = getMessageContainer({
      messageId,
      credentials: this._getCredentials(),
      requestType: 'lookupToken',
      dataElementName: 'Token',
      payload: { tokenValue },
    })
    return this._post('token', payload)
  }

  /**
   * @param {String} messageId The `messageId` is a reference for the xml request. If you had a server internally you could store all your xml request and this would be a way to locate the request if a payment was to fail. A request will still be sent if they have the same `messageId` and will be treated as a new xml request.
   * @param {String} tokenValue The token value that represents a stored card within NAB
   * @returns {Promise} {MessageInfo, RequestType, MerchantInfo, Status, Data}
   */
  deleteToken(messageId, tokenValue) {
    const payload = getMessageContainer({
      messageId,
      credentials: this._getCredentials(),
      requestType: 'deleteToken',
      dataElementName: 'Token',
      payload: {
        actionType: 'delete',
        tokenValue,
      },
    })
    return this._post('token', payload)
  }

  /**
   * @param {String} messageId The `messageId` is a reference for the xml request. If you had a server internally you could store all your xml request and this would be a way to locate the request if a payment was to fail. A request will still be sent if they have the same `messageId` and will be treated as a new xml request.
   * @param {Object} paymentDetails Payment Details
   * @param {String} paymentDetails.crn Unique identifier of customer or token.
   * @param {String} paymentDetails.currency Default currency is "AUD" â€“ Australian Dollars.
   * @param {String} paymentDetails.amount Transaction amount in cents.
   * @param {String} paymentDetails.transactionReference The transaction identifier. This value will appear against all processed transactions. Typically an invoice number. E.g. "invoice12345".
   * @returns {Promise} {MessageInfo, RequestType, MerchantInfo, Status, Data}
   */
  async triggerPayment(messageId, paymentDetails) {
    const payload = getMessageContainer({
      messageId,
      credentials: this._getCredentials(),
      requestType: 'Periodic',
      dataElementName: 'Periodic',
      payload: {
        currency: 'AUD',
        ...paymentDetails,
        actionType: 'trigger',
        periodicType: '8',
      },
    })
    const response = await this._post('periodic', payload)

    /**
     * The test API returns inconsistently formatted timestamps.
     * The token CRUD endpoints return an ISO 8601 formatted `messageTimestamp`,
     * for example: `2018-07-18T10:27:38.738Z`.
     * The trigger payment endpoint, however, returns a non ISO formatted
     * `messageTimestamp`, for example: `20181807202744603000+600`.
     * Unfortunately the NAB transact documentation is inconsistent here again.
     */
    response.MessageInfo.messageTimestamp = normaliseDateString(
      response.MessageInfo.messageTimestamp
    )
    return response
  }
}
