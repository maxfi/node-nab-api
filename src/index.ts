import { getMessageContainer } from './utils/get-message-container';
import { normaliseDateString } from './utils/normalise-date-string';
import { sendRequest } from './utils/send-request';

interface ResponseBase {
  MessageInfo: MessageInfo;
  RequestType: string;
  MerchantInfo: MerchantInfo;
  Status: Status;
}

interface NABTransactTokenResponse extends ResponseBase {
  RequestType: 'Payment' | 'lookupToken' | 'deleteToken';
  Data: Data;
}

interface NABTransactPaymentResponse extends ResponseBase {
  RequestType: 'Periodic';
  Data: TriggerPaymentData;
}

interface TriggerPaymentData {
  crn: string;
  actionType: string;
  periodicType: string;
  responseCode: string;
  responseText: string;
  successful: string;
  transactionReference: string;
  amount: string;
  currency: string;
  txnID: string;
  settlementDate: string;
  CreditCardInfo: CreditCardInfo;
}

interface Data {
  responseCode: string;
  responseText: string;
  successful: string;
  tokenValue: string;
  CreditCardInfo: CreditCardInfo;
  amount: string;
  transactionReference: string;
}

interface CreditCardInfo {
  pan: string;
  expiryDate: string;
  cardType: string;
  cardDescription: string;
}

interface MerchantInfo {
  merchantID: string;
}

interface MessageInfo {
  messageID: string;
  messageTimestamp: string;
  apiVersion: string;
}

interface Status {
  statusCode: string;
  statusDescription: string;
}

/**
 * Extracts the relevant data from the response
 * @private
 * @param response - API response as Object
 * @param  el - Name of the root XML element that contains the data
 */
const clean = (response: any, el: string) => {
  const result = response.NABTransactMessage;
  result.Data = result[el][el + 'List'][el + 'Item'];
  delete result.Data._attributes;
  delete result[el];
  return result;
};

export class NAB {
  private merchantId: string;
  private password: string;
  private baseUrl: string;
  private timeout?: string;
  /**
   * Creates an instance of the NAB class.
   * @constructor
   * @param config Config.
   * @param config.merchantId 5 character merchant ID supplied by NAB Transact.
   * @param config.password Transaction password. Password used for authentication of the merchant's request message, supplied by NAB Transact.
   * @param options Options.
   * @param options.testMode Set to `true` to direct requests to the NAB payment gateway test environment.
   * @param options.timeout Timeout value used, in seconds.
   */
  constructor(config: {
    merchantId: string;
    password: string;
    testMode: boolean;
    timeout?: string;
  }) {
    this.merchantId = config.merchantId;
    this.password = config.password;
    this.baseUrl = config.testMode
      ? 'https://demo.transact.nab.com.au/xmlapi/'
      : 'https://transact.nab.com.au/live/xmlapi/';
    this.timeout = config.timeout;
  }

  private _getCredentials() {
    return {
      merchantId: this.merchantId,
      password: this.password,
    };
  }

  private async _post(
    type: 'token',
    payload: object
  ): Promise<NABTransactTokenResponse>;
  private async _post(
    type: 'periodic',
    payload: object
  ): Promise<NABTransactPaymentResponse>;
  private async _post(type: 'token' | 'periodic', payload: object) {
    // Capitalize the element. Eg -> 'token' to 'Token'
    const el = type.charAt(0).toUpperCase() + type.slice(1);
    const url = this.baseUrl + type;
    const response = await sendRequest(url, payload);
    return clean(response, el);
  }

  /**
   * @summary Add a new token
   * @param messageId The `messageId` is a reference for the xml request. If you had a server internally you could store all your xml request and this would be a way to locate the request if a payment was to fail. A request will still be sent if they have the same `messageId` and will be treated as a new xml request.
   * @param tokenDetails Token details.
   * @param tokenDetails.cardNumber Credit card number.
   * @param tokenDetails.expiryDate Credit card expiry date.
   */
  public addToken(
    messageId: string,
    tokenDetails: { cardNumber: string; expiryDate: string }
  ) {
    const payload = getMessageContainer({
      messageId,
      credentials: this._getCredentials(),
      requestType: 'addToken',
      dataElementName: 'Token',
      payload: {
        ...tokenDetails,
        tokenType: 1, // The type of token created. Defaults to 1 if absent. Type 1 is 16 digits, not based on the card number, failing the LUHN check.
      },
      options: { timeout: this.timeout },
    });
    return this._post('token', payload);
  }

  /**
   * @param messageId The `messageId` is a reference for the xml request. If you had a server internally you could store all your xml request and this would be a way to locate the request if a payment was to fail. A request will still be sent if they have the same `messageId` and will be treated as a new xml request.
   * @param tokenValue The token value that represents a stored card within NAB Transact.
   */
  public lookupToken(messageId: string, tokenValue: string) {
    const payload = getMessageContainer({
      messageId,
      credentials: this._getCredentials(),
      requestType: 'lookupToken',
      dataElementName: 'Token',
      payload: { tokenValue },
      options: { timeout: this.timeout },
    });
    return this._post('token', payload);
  }

  /**
   * @param messageId The `messageId` is a reference for the xml request. If you had a server internally you could store all your xml request and this would be a way to locate the request if a payment was to fail. A request will still be sent if they have the same `messageId` and will be treated as a new xml request.
   * @param tokenValue The token value that represents a stored card within NAB
   */
  public deleteToken(messageId: string, tokenValue: string) {
    const payload = getMessageContainer({
      messageId,
      credentials: this._getCredentials(),
      requestType: 'deleteToken',
      dataElementName: 'Token',
      payload: {
        actionType: 'delete',
        tokenValue,
      },
      options: { timeout: this.timeout },
    });
    return this._post('token', payload);
  }

  /**
   * @param messageId The `messageId` is a reference for the xml request. If you had a server internally you could store all your xml request and this would be a way to locate the request if a payment was to fail. A request will still be sent if they have the same `messageId` and will be treated as a new xml request.
   * @param paymentDetails Payment Details
   * @param paymentDetails.crn Unique identifier of customer or token.
   * @param paymentDetails.currency Default currency is "AUD" â€“ Australian Dollars.
   * @param paymentDetails.amount Transaction amount in cents.
   * @param paymentDetails.transactionReference The transaction identifier. This value will appear against all processed transactions. Typically an invoice number. E.g. "invoice12345".
   */
  public async triggerPayment(
    messageId: string,
    paymentDetails: {
      crn: string;
      amount: string;
      transactionReference: string;
    }
  ) {
    const payload = getMessageContainer({
      messageId,
      credentials: this._getCredentials(),
      requestType: 'Periodic',
      dataElementName: 'Periodic',
      payload: {
        currency: 'AUD',
        ...paymentDetails,
        actionType: 'trigger',
        periodicType: '8',
      },
      options: { timeout: this.timeout },
    });
    const response = await this._post('periodic', payload);

    /**
     * The test API returns inconsistently formatted timestamps.
     * The token CRUD endpoints return an ISO 8601 formatted `messageTimestamp`,
     * for example: `2018-07-18T10:27:38.738Z`.
     * The trigger payment endpoint, however, returns a non ISO formatted
     * `messageTimestamp`, for example: `20181807202744603000+600`.
     * Unfortunately the NAB transact documentation is inconsistent here again.
     */
    response.MessageInfo.messageTimestamp = normaliseDateString(
      response.MessageInfo.messageTimestamp
    );

    // Reference: XMLAPIIntegrationGuide.pdf - 4.10 Appendix J: NAB Transact Bank Response Codes
    const isApproved = ['00', '08', '11', '16'].includes(
      response.Data.responseCode
    );

    return { ...response, isApproved };
  }
}
