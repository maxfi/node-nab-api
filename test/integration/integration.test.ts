import { NAB } from '../../src';
import { isoDateStringRegex } from '../helpers/iso-date-string-regex';

const nab = new NAB({
  merchantId: 'XYZ0010',
  password: 'abcd1234',
  testMode: true,
});

const messageTimeStampPropertyMatcher = {
  MessageInfo: {
    messageTimestamp: expect.stringMatching(isoDateStringRegex),
  },
};

/**
 * NOTE: From docs (XML_API_Guide_Tokenisation.pdf)
 * > One card number will always generate the same Token.
 * However, this doesn't appear to be case when a token is created,
 * then deleted, and created again.
 */

test('addToken', async () => {
  const result = await nab.addToken('123456789', {
    cardNumber: '4219448174236125',
    expiryDate: '01/26',
  });

  expect(result).toMatchSnapshot(messageTimeStampPropertyMatcher);
});

test('lookupToken', async () => {
  const {
    Data: { tokenValue },
  } = await nab.addToken('123456789', {
    cardNumber: '4549771486564913',
    expiryDate: '09/22',
  });

  const result = await nab.lookupToken('123456789', tokenValue);

  expect(result).toMatchSnapshot({
    ...messageTimeStampPropertyMatcher,
    Data: { tokenValue: expect.stringMatching(tokenValue) },
  });
});

test('deleteToken', async () => {
  const {
    Data: { tokenValue },
  } = await nab.addToken('123456789', {
    cardNumber: '4334899786622765',
    expiryDate: '08/27',
  });

  const deleteResult = await nab.deleteToken('123456789', tokenValue);
  expect(deleteResult).toMatchSnapshot(messageTimeStampPropertyMatcher);

  const lookupResult = await nab.lookupToken('123456789', tokenValue);
  expect(lookupResult).toMatchSnapshot(messageTimeStampPropertyMatcher);
});

test('triggerPayment - success', async () => {
  const result = await nab.triggerPayment('123456789', {
    crn: 'testTokenId180703',
    amount: '100',
    transactionReference: 'test transaction - node-nab-api',
  });

  expect(result).toMatchSnapshot({
    ...messageTimeStampPropertyMatcher,
    Data: {
      txnID: expect.stringMatching(/^\d+$/),
      settlementDate: expect.stringMatching(/^\d{8}$/),
    },
  });
});

test('triggerPayment - failure', async () => {
  const result = await nab.triggerPayment('123456789', {
    crn: 'testTokenId180703',
    amount: '105',
    transactionReference: 'test transaction - node-nab-api',
  });

  expect(result).toMatchSnapshot({
    ...messageTimeStampPropertyMatcher,
    Data: {
      txnID: expect.stringMatching(/^\d+$/),
      settlementDate: expect.stringMatching(/^\d{8}$/),
    },
  });
});
