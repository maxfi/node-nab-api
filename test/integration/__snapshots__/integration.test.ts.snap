// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addToken 1`] = `
Object {
  "Data": Object {
    "CreditCardInfo": Object {
      "cardDescription": "Visa",
      "cardType": "6",
      "expiryDate": "01/26",
      "pan": "421944XXXXXXX125",
    },
    "amount": "100",
    "responseCode": "00",
    "responseText": "Successful",
    "successful": "yes",
    "tokenValue": "3424780602475871",
    "transactionReference": Object {},
  },
  "MerchantInfo": Object {
    "merchantID": "XYZ0010",
  },
  "MessageInfo": Object {
    "apiVersion": "xml-4.2",
    "messageID": "123456789",
    "messageTimestamp": StringMatching /\\^\\\\d\\{4\\}-\\[01\\]\\\\d-\\[0-3\\]\\\\dT\\[0-2\\]\\\\d:\\[0-5\\]\\\\d:\\[0-5\\]\\\\d\\\\\\.\\\\d\\+\\(\\[\\+-\\]\\[0-2\\]\\\\d:\\[0-5\\]\\\\d\\|Z\\)\\$/u,
  },
  "RequestType": "Payment",
  "Status": Object {
    "statusCode": "0",
    "statusDescription": "Normal",
  },
}
`;

exports[`deleteToken 1`] = `
Object {
  "Data": Object {
    "CreditCardInfo": Object {
      "cardDescription": Object {},
      "cardType": "0",
      "expiryDate": Object {},
      "pan": Object {},
    },
    "amount": "0",
    "responseCode": "00",
    "responseText": "Successful",
    "successful": "yes",
    "tokenValue": "5966584544780748",
    "transactionReference": Object {},
  },
  "MerchantInfo": Object {
    "merchantID": "XYZ0010",
  },
  "MessageInfo": Object {
    "apiVersion": "xml-4.2",
    "messageID": "123456789",
    "messageTimestamp": StringMatching /\\^\\\\d\\{4\\}-\\[01\\]\\\\d-\\[0-3\\]\\\\dT\\[0-2\\]\\\\d:\\[0-5\\]\\\\d:\\[0-5\\]\\\\d\\\\\\.\\\\d\\+\\(\\[\\+-\\]\\[0-2\\]\\\\d:\\[0-5\\]\\\\d\\|Z\\)\\$/u,
  },
  "RequestType": "Payment",
  "Status": Object {
    "statusCode": "0",
    "statusDescription": "Normal",
  },
}
`;

exports[`deleteToken 2`] = `
Object {
  "Data": Object {
    "CreditCardInfo": Object {
      "cardDescription": Object {},
      "cardType": "0",
      "expiryDate": Object {},
      "pan": Object {},
    },
    "amount": "0",
    "responseCode": "701",
    "responseText": "Token does not exist",
    "successful": "yes",
    "tokenValue": "5966584544780748",
    "transactionReference": Object {},
  },
  "MerchantInfo": Object {
    "merchantID": "XYZ0010",
  },
  "MessageInfo": Object {
    "apiVersion": "xml-4.2",
    "messageID": "123456789",
    "messageTimestamp": StringMatching /\\^\\\\d\\{4\\}-\\[01\\]\\\\d-\\[0-3\\]\\\\dT\\[0-2\\]\\\\d:\\[0-5\\]\\\\d:\\[0-5\\]\\\\d\\\\\\.\\\\d\\+\\(\\[\\+-\\]\\[0-2\\]\\\\d:\\[0-5\\]\\\\d\\|Z\\)\\$/u,
  },
  "RequestType": "Payment",
  "Status": Object {
    "statusCode": "0",
    "statusDescription": "Normal",
  },
}
`;

exports[`lookupToken 1`] = `
Object {
  "Data": Object {
    "CreditCardInfo": Object {
      "cardDescription": "Visa",
      "cardType": "6",
      "expiryDate": "09/22",
      "pan": "454977XXXXXXX913",
    },
    "amount": "100",
    "responseCode": "00",
    "responseText": "Successful",
    "successful": "yes",
    "tokenValue": StringMatching /7151059647617577/,
    "transactionReference": Object {},
  },
  "MerchantInfo": Object {
    "merchantID": "XYZ0010",
  },
  "MessageInfo": Object {
    "apiVersion": "xml-4.2",
    "messageID": "123456789",
    "messageTimestamp": StringMatching /\\^\\\\d\\{4\\}-\\[01\\]\\\\d-\\[0-3\\]\\\\dT\\[0-2\\]\\\\d:\\[0-5\\]\\\\d:\\[0-5\\]\\\\d\\\\\\.\\\\d\\+\\(\\[\\+-\\]\\[0-2\\]\\\\d:\\[0-5\\]\\\\d\\|Z\\)\\$/u,
  },
  "RequestType": "Payment",
  "Status": Object {
    "statusCode": "0",
    "statusDescription": "Normal",
  },
}
`;

exports[`triggerPayment - failure 1`] = `
Object {
  "Data": Object {
    "CreditCardInfo": Object {
      "cardDescription": "Master Card",
      "cardType": "5",
      "expiryDate": "12/23",
      "pan": "555555...444",
    },
    "actionType": "trigger",
    "amount": "105",
    "crn": "testTokenId180703",
    "currency": "AUD",
    "periodicType": "8",
    "responseCode": "05",
    "responseText": "Do Not Honour",
    "settlementDate": StringMatching /\\^\\\\d\\{8\\}\\$/,
    "successful": "no",
    "transactionReference": "test transaction - node-nab-api",
    "txnID": StringMatching /\\^\\\\d\\+\\$/,
  },
  "MerchantInfo": Object {
    "merchantID": "XYZ0010",
  },
  "MessageInfo": Object {
    "apiVersion": "spxml-3.0",
    "messageID": "123456789",
    "messageTimestamp": StringMatching /\\^\\\\d\\{4\\}-\\[01\\]\\\\d-\\[0-3\\]\\\\dT\\[0-2\\]\\\\d:\\[0-5\\]\\\\d:\\[0-5\\]\\\\d\\\\\\.\\\\d\\+\\(\\[\\+-\\]\\[0-2\\]\\\\d:\\[0-5\\]\\\\d\\|Z\\)\\$/u,
  },
  "RequestType": "Periodic",
  "Status": Object {
    "statusCode": "0",
    "statusDescription": "Normal",
  },
  "isApproved": false,
}
`;

exports[`triggerPayment - success 1`] = `
Object {
  "Data": Object {
    "CreditCardInfo": Object {
      "cardDescription": "Master Card",
      "cardType": "5",
      "expiryDate": "12/23",
      "pan": "555555...444",
    },
    "actionType": "trigger",
    "amount": "100",
    "crn": "testTokenId180703",
    "currency": "AUD",
    "periodicType": "8",
    "responseCode": "00",
    "responseText": "Approved",
    "settlementDate": StringMatching /\\^\\\\d\\{8\\}\\$/,
    "successful": "yes",
    "transactionReference": "test transaction - node-nab-api",
    "txnID": StringMatching /\\^\\\\d\\+\\$/,
  },
  "MerchantInfo": Object {
    "merchantID": "XYZ0010",
  },
  "MessageInfo": Object {
    "apiVersion": "spxml-3.0",
    "messageID": "123456789",
    "messageTimestamp": StringMatching /\\^\\\\d\\{4\\}-\\[01\\]\\\\d-\\[0-3\\]\\\\dT\\[0-2\\]\\\\d:\\[0-5\\]\\\\d:\\[0-5\\]\\\\d\\\\\\.\\\\d\\+\\(\\[\\+-\\]\\[0-2\\]\\\\d:\\[0-5\\]\\\\d\\|Z\\)\\$/u,
  },
  "RequestType": "Periodic",
  "Status": Object {
    "statusCode": "0",
    "statusDescription": "Normal",
  },
  "isApproved": true,
}
`;
